server.ts

```diff
@@ -80,28 +85,33 @@ const getsurvey = async (query: string | ParsedQs, req: Request<{}>, res: Respon
 app.get("/s/", async (req, res) => {
   //For debugging purposes. Prints encrypted version of url. In the future, this will be done in researcher menu.
   const cipher = crypto.createCipheriv(crypto_algorithm, private_key_example, iv_example);
-  let encrypted = cipher.update(JSON.stringify(req.query), 'utf8', 'hex');
-  console.log(encrypted += cipher.final('hex'));
-
+  let encrypted = cipher.update(JSON.stringify(req.query), "utf8", "hex");
+  console.log(encrypted += cipher.final("hex"));
   console.log(req.query);
   getsurvey(req.query, req, res);
 });
 
-app.get('/e/:data', async (req, res) => {
+app.get("/e/:data", async (req, res) => {
   //in the future, private_key and iv will be obtained through researcher database
   try {
     const decipher = crypto.createDecipheriv(crypto_algorithm, private_key_example, iv_example);
-    let decrypted = decipher.update(req.params.data, 'hex', 'utf8');
-    decrypted += decipher.final('utf8');
+    let decrypted = decipher.update(req.params.data, "hex", "utf8");
+    decrypted += decipher.final("utf8");
     getsurvey(JSON.parse(decrypted), req, res);
   } catch (error) {
     console.error(error);
-    res.redirect('/');
+    res.redirect("/");
   }
 });
 
-app.post("/survey", (req, res) => {
+app.post("/survey", csrfProtection, (req, res) => {
```

Comment:
Running locally I am getting an error on submission, which I think is due to browser privacy protections (thought I am seeing a similar error in both Safari and Chrome).

```
ForbiddenError: invalid csrf token
    at csrf (/Users/mark/Documents/Academia/Penn/Surveyor/node_modules/csurf/index.js:112:19)
    at Layer.handle [as handle_request] (/Users/mark/Documents/Academia/Penn/Surveyor/node_modules/express/lib/router/layer.js:95:5)
    at next (/Users/mark/Documents/Academia/Penn/Surveyor/node_modules/express/lib/router/route.js:137:13)
    at Route.dispatch (/Users/mark/Documents/Academia/Penn/Surveyor/node_modules/express/lib/router/route.js:112:3)
    at Layer.handle [as handle_request] (/Users/mark/Documents/Academia/Penn/Surveyor/node_modules/express/lib/router/layer.js:95:5)
    at /Users/mark/Documents/Academia/Penn/Surveyor/node_modules/express/lib/router/index.js:281:22
    at Function.process_params (/Users/mark/Documents/Academia/Penn/Surveyor/node_modules/express/lib/router/index.js:335:12)
    at next (/Users/mark/Documents/Academia/Penn/Surveyor/node_modules/express/lib/router/index.js:275:10)
    at /Users/mark/Documents/Academia/Penn/Surveyor/node_modules/body-parser/lib/read.js:130:5
    at invokeCallback (/Users/mark/Documents/Academia/Penn/Surveyor/node_modules/raw-body/index.js:224:16)
```

Any thoughts on how to deal with this? 

Notably I think cookies are a bit less robust as a solution now days because of of these kinds of protections. This might be an argument for piping the variable through the submission, or having a unique token in the initial survey response that we use to maintain state?

Reply:
Huh, I didn't get that error last night. I'll look into it this afternoon. 
I think having the time stamp run on the client side might be a better solution. I'll try something out and update you.

Reply:
It works now. Lmk if does for you. I got rid of the cookies and went with the strategy below.

I currently am setting time stamp first by injecting the start time through PUG input as you suggested, and then when the data hits the endpoint /survey, it timestamps the recording. 

I explored trying to get timestamp as soon as somebody clicks the submit button, it not very easy in since we are server side rendering PUG. It would be super easy in React-- a client side front end framework, where we could start the time as soon as the person is active (i.e interacts with the first question) through submit.

Reply:
Agreed re React. I'm not against doing something like that. When you get a chance, start an issue around the pros and cons of using React for this?